goalie/interpolation.py

```diff
@@ -42,108 +55,113 @@ def project(source, target_space, **kwargs):
+    """
+    return transfer(source, target_space, transfer_method="interpolate", **kwargs)
 
-    This function extends :func:`firedrake.projection.project`` to account for mixed
-    spaces.
 
-    :arg source: the Function to be projected
+@PETSc.Log.EventDecorator()
+def project(source, target_space, **kwargs):
+    """
+    A wrapper for :func:`transfer` with ``transfer_method="interpolate"``.
+    """
+    return transfer(source, target_space, transfer_method="project", **kwargs)
+
+
+@PETSc.Log.EventDecorator()
+def _transfer_forward(source, target, transfer_method, **kwargs):
+    """
+    Apply mesh-to-mesh transfer operator to a Function.
+
+    This function extends the functionality of :func:`firedrake.__future__.interpolate`
+    and :func:`firedrake.projection.project` to account for mixed spaces.
+
+    :arg source: the Function to be transferred
     :type source: :class:`firedrake.function.Function`
-    :arg target: the Function which we seek to project onto
-    :type target: :class:`firedrake.function.Function`.
-    :returns: the target projection
+    :arg target: the Function which we seek to transfer onto
+    :type target: :class:`firedrake.function.Function`
+    :kwarg transfer_method: the method to use for the transfer. Options are
+        "interpolate" (default) and "project".
+    :type transfer_method: str
+    :returns: the transferred Function
     :rtype: :class:`firedrake.function.Function`
 
-    Extra keyword arguments are passed to :func:`firedrake.projection.project``.
+    Extra keyword arguments are passed to :func:`firedrake.__future__.interpolate` or
+        :func:`firedrake.projection.project`.
     """
     Vs = source.function_space()
     Vt = target.function_space()
-    if hasattr(Vs, "num_sub_spaces"):
-        if not hasattr(Vt, "num_sub_spaces"):
-            raise ValueError(
-                "Source space has multiple components but target space does not."
-            )
-        if Vs.num_sub_spaces() != Vt.num_sub_spaces():
-            raise ValueError(
-                "Inconsistent numbers of components in source and target spaces:"
-                f" {Vs.num_sub_spaces()} vs. {Vt.num_sub_spaces()}."
-            )
-    elif hasattr(Vt, "num_sub_spaces"):
-        raise ValueError(
-            "Target space has multiple components but source space does not."
-        )
+    _validate_matching_spaces(Vs, Vt)
     assert isinstance(target, firedrake.Function)
     if hasattr(Vt, "num_sub_spaces"):
         for s, t in zip(source.subfunctions, target.subfunctions):
-            t.project(s, **kwargs)
+            if transfer_method == "interpolate":
+                t.interpolate(s, **kwargs)
+            else:
+                t.project(s, **kwargs)
     else:
-        target.project(source, **kwargs)
+        if transfer_method == "interpolate":
+            target.interpolate(source, **kwargs)
+        else:
+            target.project(source, **kwargs)
```

Comment:
Instead of using `else` it would make sense to use `elif transfer_method == "project"` and then to add a separate `else` case which returns an error that the requested transfer method was not recognised. You should also add unit tests for this.

Reply:
Makes sense, sorry. Fixed in 98b40fea0dbb5e26ed49d935f086f168b1001c73